x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import dateMath from '@elastic/datemath';
+import { i18n } from '@kbn/i18n';
+
+import { ContextMenuAction, ContextMenuActionsRegistryProvider } from 'ui/embeddable';
+import { PanelActionAPI } from 'ui/embeddable/context_menu_actions/types';
+import { kfetch } from 'ui/kfetch';
+import { toastNotifications } from 'ui/notify';
+import { API_BASE_URL_V1 } from '../../common/constants';
+
+const API_BASE_URL = `${API_BASE_URL_V1}/generate/immediate/csv/saved-object/`;
+
+class GetCsvReportPanelAction extends ContextMenuAction {
+  constructor() {
+    super(
+      {
+        displayName: i18n.translate('xpack.reporting.dashboard.downloadCsvPanelTitle', {
+          defaultMessage: 'Download CSV',
+        }),
+        id: 'downloadCsvReport',
+        parentPanelId: 'mainMenu',
+      },
+      {
+        icon: 'document',
+      }
+    );
+  }
+
+  public async generateJobParams({ searchEmbeddable }: { searchEmbeddable: any }) {
+    const adapters = searchEmbeddable.getInspectorAdapters();
+    if (!adapters) {
+      return {};
+    }
+
+    if (adapters.requests.requests.length === 0) {
+      return {};
+    }
+
+    return searchEmbeddable.searchScope.searchSource.getSearchRequestBody();
+  }
+
+  // @TODO: Clean this up once we update SavedSearch's interface
+  // and location in the file-system. `viewMode` also has an enum
+  // buried inside of the dashboard folder we could use vs a bare string
+  public isVisible = (panelActionAPI: PanelActionAPI): boolean => {
+    const { embeddable, containerState } = panelActionAPI;
+
+    return (
+      containerState.viewMode !== 'edit' && !!embeddable && embeddable.hasOwnProperty('savedSearch')
+    );
+  };
+
+  // TODO: Need to expose some of the interface here from SavedSearch
+  // as well as pass things like columns into the API call
+  public onClick = async (panelActionAPI: PanelActionAPI) => {
+    const { embeddable } = panelActionAPI as any;
+    const {
+      timeRange: { from, to },
+    } = embeddable;
+
+    if (!embeddable) {
+      return;
+    }
+
+    const searchEmbeddable = embeddable;
+    const state = await this.generateJobParams({ searchEmbeddable });
+
+    const id = `search:${embeddable.savedSearch.id}`;
+    const filename = embeddable.savedSearch.title;
+    const timezone = embeddable.$rootScope.chrome.getUiSettingsClient().get('dateFormat:tz');
+    const fromTime = dateMath.parse(from);
+    const toTime = dateMath.parse(to);
+
+    if (!fromTime || !toTime) {
+      return this.onGenerationFail(
+        new Error(`Invalid time range: From: ${fromTime}, To: ${toTime}`)
+      );
+    }
+
+    const body = JSON.stringify({
+      timerange: {
+        min: fromTime.valueOf(),
+        max: toTime.valueOf(),
```

Comment:
These should be formatted as string, so if the time is before the epoch, it is not a negative value.

Elasticsearch will fail for this type of date:
```
  "timerange": {
    "max": -1893382451814,
    "min": -2177423570985,
    "timezone": "UTC"
  }
```

It is fine with this:
```
  "timerange": {
    "max": "1911-01-01T00:00:00.000Z",
    "min": "1901-01-01T00:00:00.000Z",
    "timezone": "UTC"
  }

```

```
  "timerange": {
    "max": "1911-01-01T00:00:00.000Z",
    "min": "1901-01-01T00:00:00.000Z",
    "timezone": "UTC"
  }
```

Strangely, somehow this is fine though: 
```
  "timerange": {
    "max": -1861920000000,
    "min": -2177452800000,
    "timezone": "UTC"
  }

```

Reply:
Done!